src/transformations/defmodel.hpp

```diff
@@ -0,0 +1,1890 @@
+            comp.displacementType() == STR_3D) {
+            if (dmmf->mHorizontalOffsetUnit.empty()) {
+                throw ParsingException("horizontal_offset_unit should be "
+                                       "defined as there is a component with "
+                                       "displacement_type = horizontal/3d");
+            }
+            if (dmmf->mHorizontalOffsetMethod.empty()) {
+                throw ParsingException("horizontal_offset_method should be "
+                                       "defined as there is a component with "
+                                       "displacement_type = horizontal/3d");
+            }
+        }
+        if (comp.displacementType() == STR_VERTICAL ||
+            comp.displacementType() == STR_3D) {
+            if (dmmf->mVerticalOffsetUnit.empty()) {
+                throw ParsingException("vertical_offset_unit should be defined "
+                                       "as there is a component with "
+                                       "displacement_type = vertical/3d");
+            }
+        }
+        if (dmmf->mHorizontalOffsetUnit == STR_DEGREE &&
+            comp.spatialModel().interpolationMethod != STR_BILINEAR) {
+            throw ParsingException("horizontal_offset_unit = degree can only "
+                                   "be used with interpolation_method = "
+                                   "bilinear");
+        }
+    }
+
+    if (dmmf->mHorizontalOffsetUnit == STR_DEGREE &&
+        dmmf->mHorizontalOffsetMethod != STR_ADDITION) {
+        throw ParsingException("horizontal_offset_unit = degree can only be "
+                               "used with horizontal_offset_method = addition");
+    }
+
+    return dmmf;
+}
+
+// ---------------------------------------------------------------------------
+
+SpatialExtent SpatialExtent::parse(const json &j) {
+    SpatialExtent ex;
+
+    const std::string type = getReqString(j, "type");
+    if (type != "bbox") {
+        throw ParsingException("unsupported type of extent");
+    }
+
+    const json jParameter = getObjectMember(j, "parameters");
+    const json jBbox = getArrayMember(jParameter, "bbox");
+    if (jBbox.size() != 4) {
+        throw ParsingException("bbox is not an array of 4 numeric elements");
+    }
+    for (int i = 0; i < 4; i++) {
+        if (!jBbox[i].is_number()) {
+            throw ParsingException(
+                "bbox is not an array of 4 numeric elements");
+        }
+    }
+    ex.mMinx = jBbox[0].get<double>();
+    ex.mMiny = jBbox[1].get<double>();
+    ex.mMaxx = jBbox[2].get<double>();
+    ex.mMaxy = jBbox[3].get<double>();
+
+    ex.mMinxRad = DegToRad(ex.mMinx);
+    ex.mMinyRad = DegToRad(ex.mMiny);
+    ex.mMaxxRad = DegToRad(ex.mMaxx);
+    ex.mMaxyRad = DegToRad(ex.mMaxy);
+
+    return ex;
+}
+
+// ---------------------------------------------------------------------------
+
+Component Component::parse(const json &j) {
```

Comment:
Move this to its own utility/class/header?

Reply:
excellent suggestion. I've moved all implementation details to a defmodel_impl.hpp file